include/ofi_atomic_queue.h

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2022 UT-Battelle ORNL. All rights reserved
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef OFI_ATOMIC_QUEUE_H
+#define OFI_ATOMIC_QUEUE_H
+
+#include <ofi_atom.h>
+
+/*
+ * This is an atomic queue, meaning no need for locking. One example usage
```

Comment:
This looks like a circular queue with atomic read/write capabilities. This allows us to remove the spin lock on the SHM region (which is a big improvement). Fundamentally, our SHM implementation is still using the wrong algorithm for optimal performance, so we can only expect to scale so much.

The two major issues are:
1. Circular Queues instead of single reader, multiple writer FIFO queues
2. Receiving process (peer) owning the SHM, leading to multiple processes polling on remote SHM, which causes lots of cache invalidations, and clogs up memory bandwidth

Don't get too attached to this queue... I plan to align our SHM approach to use Nemesis algorithm, which means this queue will have to go.


Background:
The Nemesis algorithm uses atomic single reader, multiple writer FIFO queues for both the receive queue, and the free queue. You can optimize this by using a LIFO free queue (for cache hits), and re-using the receive queue to return free queue entries to the sender.  The [ Nemesis paper ](https://digital.library.unt.edu/ark:/67531/metadc873149/m2/1/high_res_d/881588.pdf) explains this reasoning well in section 2.1.1

Reply:
@a-szegel - I agree that the nemesis algorithm looks to be ideal for this use case, only their naming sucks.

How quickly do you think you would have an implementation available?  This PR has other optimizations worth pulling in, but it may not be worth updating the mpmc queue patch if it will simply be replaced.

Reply:
I think we should pull this in since it is ready, and it could be a few weeks till I have something.  Still a beginner with libfabric :)

Reply:
@a-szegel, will the single reader approach work for multiple progress thread scenario?

Reply:
You still need locks for multiple threads.  The implementation you put up for the atomic circular queue does have to be multi-reader, and multi-writer safe (as written).  I thought you had already switched us to Nemesis which is why I made the comment that we only need a single reader.

Reply:
open MPI separates the multiple thread case from the single thread case b/c each uncontested pthread mutex lock/unlock takes ~2 nanoseconds.

Reply:
Agreed - multiple threads on the reader side need to be serialized for message buffer matching to behave correctly.  So, we'll have a higher level progress lock already.

Reply:
ok, that's the last patch in this PR.

Reply:
@shefty, what's your plan with this PR?

Reply:
I was waiting for your updates.  I haven't reviewed it yet.

Reply:
still thinking on the potential advantage of the nemesis algorithm. Since we're going to have a common lock which all threads need to take before reading from the queue in the multi-thread case, how will that impact performance? The current lock which we use should be a no-op for the single thread case, so I guess that should be more efficient. But it would be interesting to quantify this performance difference between the two approaches.

Reply:
The nemesis algorithm should have better performance with the many writers (processes, not threads) to one reader case.  It should be able to scale a lot better than the circular queue for large instances under heavy load.  We will still need fast boxes to get ~250ns pt2pt latency for small messages but that can come later.  We will make some performance graphs and we won't take it if it isn't better :).

Reply:
One of the areas which I'm interested in is matching the on node performance of OMPI for small message sizes with SHM provider. For larger messages above 4K, I currently show that both OMPI and SHM yield about the same performance. MPICH, OMPI and other middleware all implement their own shared memory mechanisms, and I think it makes a lot of sense if SHM provider can be used for these middleware software if they use libfabric. But in order to get there, we need to match their performance. The question I'm trying to answer by quantifying the nemesis algorithm, is if that's the missing piece to get us to match the performance in OMPI, or if there are other areas of improvement we need to look at.